web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,400 @@
+   * @param d Holds one interval's data on the timeline.
+   */
+  private getTooltipContent = (d: Item) => {
+    const duration = (d.endTime - d.startTime);
+    const options = ({
+      round: true
+    } as HumanizeDurationOptions)
+    const output = TimelineComponent.HUMANIZER.humanize(duration, options);
+
+    // HTML representation of all required data.
+    return `<b>Push ID: ${d.pushID.slice(d.pushID.indexOf('@') + 1)}</b>
+      <br/>
+      <b>Final State: ${d.state}</b>
+      <br/>
+      <b>Start Time: ${this.formatDate(d.startTime)}</b>
+      <br/>
+      <b> End Time: ${this.formatDate(d.endTime)}</b>
+      <br/>
+      <b>Duration: ${output}</b>
+      `
+  };
+
+  /**
+   * Create tooltip to display each interval's content.
+   * @param el Encasing element that holds the tooltip
+   */
+  private createTooltip = (el: any) => { // TODO: tachyons?
+    el.style('position', 'absolute')
+      .style('pointer-events', 'none')
+      .style('top', 0)
+      .style('opacity', 0)
+      .style('background', 'white')
+      .style('border-radius', '5px')
+      .style('box-shadow', '0 0 10px rgba(0,0,0,.25)')
+      .style('padding', '10px')
+      .style('line-height', '1.3')
+      .style('font', '11px sans-serif');
+  };
+
+  /** 
+   * Defines the zoom behavior, limiting the scale with which we can zoom in and out
+   * and restricting zoom to the x-axis only. Upon zoom, the x-axis will rescale,
+   * as will the timeline intervals.
+   */
+  private zoom = d3.zoom()
+    .scaleExtent([0.75, 1000]) // Limit zoom out.
+    .translateExtent([[-100000, 0], [100000, 0]]) // Avoid scrolling too far.
+    .on('zoom', () => {
+      const transform = d3.event.transform;
+
+      const updatedScale = transform.rescaleX(this.x);
+
+      // Redraw the x-axis.
+      this.xAxis = d3
+        .axisBottom(updatedScale)
+        .tickSize(-this.height - 6)
+        .tickPadding(10)
+        .tickFormat(this.formatDate);
+
+      this.svg.select('.x.axis')
+        .call(this.xAxis)
+        .selectAll('line')
+        .style('stroke', TimelineComponent.COLOR_LIGHT_GRAY);
+
+      this.svg.selectAll('rect.interval')
+        .attr('x', (d: Item) => updatedScale(d.startTime))
+        .attr('width', (d: Item) => updatedScale(d.endTime) - updatedScale(d.startTime));
+    });
+
+  /**
+   * Creates a scrollable timeline with bars representing the duration of
+   * pushes. Every time this function is called, the previous timeline SVG is
+   * removed and a new one propagated to the screen. Timeline features include 
+   * scrolling horizontally on the x-axis, zooming in and out along the x-axis,
+   * and a tooltip display when hovering over intervals. Each interval is 
+   * color-coded according to its final state.
+   * @param pushInfos Holds all pushes for one push def.
+   */
+  private createTimeline(): void {
+    // Filter the data by adding raw protocol buffer data into Item
```

Comment:
I don't know if "raw protocol buffer data" is correct because technically the raw data are the binary files